plugin/aws/s3Bucket.go

```diff
@@ -46,23 +47,27 @@ func listObjects(ctx context.Context, client *s3Client.S3, bucket string, prefix
 	)
 
 	for i, p := range resp.CommonPrefixes {
-		prefix := awsSDK.StringValue(p.Prefix)
-		// Skip the top-level '/' prefix, it would be redundant to list it.
-		if prefix == "/" {
-			continue
+		commonPrefix := awsSDK.StringValue(p.Prefix)
+		name := strings.TrimPrefix(commonPrefix, prefix)
+		if name != "/" {
+			name = strings.TrimSuffix(name, "/")
 		}
-		entries[i] = newS3ObjectPrefix(bucket, prefix, client)
+		entries[i] = newS3ObjectPrefix(name, bucket, commonPrefix, client)
 	}
 
-	for i, o := range resp.Contents {
-		entries[numPrefixes+i] = newS3Object(
-			bucket,
-			awsSDK.StringValue(o.Key),
-			client,
-		)
+	i := numPrefixes
+	for _, o := range resp.Contents {
+		key := awsSDK.StringValue(o.Key)
+		name := strings.TrimPrefix(key, prefix)
+		if name == "" {
+			// Key has a trailing "/", so skip it
+			continue
+		}
+		entries[i] = newS3Object(name, bucket, key, client)
+		i++
 	}
 
-	return entries, nil
+	return entries[0:i], nil
```

Comment:
This could be simpler by creating `entries = make([]plugin.Entry, 0, numPrefixes+numObjects)` and using `entries = append(entries, ...)` everywhere. Then you don't have to track index.